// @rowno/sparkline/commonjs/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _d3Shape = require("d3-shape");

var _d3Scale = require("d3-scale");

var _d3Array = require("d3-array");

var curveFunction = _d3Shape.curveMonotoneX;
var defaultColors = {
  area: 'rgba(199, 228, 255, 0.5)',
  line: '#004585'
};

var Sparkline =
/*#__PURE__*/
function (_PureComponent) {
  (0, _inherits2["default"])(Sparkline, _PureComponent);

  function Sparkline() {
    (0, _classCallCheck2["default"])(this, Sparkline);
    return (0, _possibleConstructorReturn2["default"])(this, (0, _getPrototypeOf2["default"])(Sparkline).apply(this, arguments));
  }

  (0, _createClass2["default"])(Sparkline, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          lines = _this$props.lines,
          width = _this$props.width,
          height = _this$props.height;
      var maxX = (0, _d3Array.max)(lines, function (s) {
        return s.values.length - 1;
      });
      var maxY = (0, _d3Array.max)(lines, function (s) {
        return (0, _d3Array.max)(s.values);
      });
      var x = (0, _d3Scale.scaleLinear)().domain([0, maxX]).range([0, width]); // Set range to 1 to make room for the line stroke

      var y = (0, _d3Scale.scaleLinear)().domain([0, maxY]).range([height, 1]);
      var areaFunction = (0, _d3Shape.area)().x(function (d, i) {
        return x(i);
      }).y0(height).y1(function (d) {
        return y(d);
      }).curve(curveFunction);
      var lineFunction = (0, _d3Shape.line)().x(function (d, i) {
        return x(i);
      }).y(function (d) {
        return y(d);
      }).curve(curveFunction);
      return _react["default"].createElement("svg", {
        width: width,
        height: height,
        viewBox: "0 0 ".concat(width, " ").concat(height)
      }, lines.map(function (line, index) {
        var colors = (0, _objectSpread2["default"])({}, defaultColors, line.colors);
        return _react["default"].createElement("g", {
          key: line.key || index
        }, _react["default"].createElement("path", {
          d: areaFunction(line.values),
          fill: colors.area
        }, line.title && _react["default"].createElement("title", null, line.title)), _react["default"].createElement("path", {
          d: lineFunction(line.values),
          stroke: colors.line,
          fill: "none"
        }));
      }));
    }
  }]);
  return Sparkline;
}(_react.PureComponent);

exports["default"] = Sparkline;
(0, _defineProperty2["default"])(Sparkline, "displayName", 'Sparkline');
(0, _defineProperty2["default"])(Sparkline, "propTypes", {
  lines: _propTypes["default"].arrayOf(_propTypes["default"].shape({
    values: _propTypes["default"].arrayOf(_propTypes["default"].number.isRequired).isRequired,
    colors: _propTypes["default"].shape({
      area: _propTypes["default"].string.isRequired,
      line: _propTypes["default"].string.isRequired
    }),
    title: _propTypes["default"].string,
    key: _propTypes["default"].any
  })).isRequired,
  width: _propTypes["default"].number.isRequired,
  height: _propTypes["default"].number.isRequired
});
